dvc/command/pipeline.py

```diff
@@ -36,31 +36,33 @@ def __build_graph(self, target, commands, outs):
         from dvc.stage import Stage
         from dvc.repo.graph import get_pipeline
 
-        stage = Stage.load(self.repo, target)
-        G = get_pipeline(self.repo.pipelines, stage)
+        target_stage = Stage.load(self.repo, target)
+        G = get_pipeline(self.repo.pipelines, target_stage)
 
         nodes = []
-        for stage in G:
+        for stage in networkx.preorder_nodes(G, target_stage):
             if commands:
                 if stage.cmd is None:
                     continue
                 nodes.append(stage.cmd)
             elif outs:
                 for out in stage.outs:
                     nodes.append(str(out))
+                for dep in stage.deps:
+                    nodes.append(str(dep))
             else:
                 nodes.append(stage.relpath)
 
         edges = []
-        for from_stage, to_stage in G.edges():
+        for from_stage, to_stage in networkx.dfs_edges(G, target_stage):
             if commands:
                 if to_stage.cmd is None:
                     continue
                 edges.append((from_stage.cmd, to_stage.cmd))
             elif outs:
-                for from_out in from_stage.outs:
-                    for to_out in to_stage.outs:
-                        edges.append((str(from_out), str(to_out)))
+                for from_dep in from_stage.deps:
+                    for from_out in from_stage.outs:
```

Comment:
So we are not using `to_stage` here anymore, looks like a bug.

Reply:
This seems to be complete. I think it's because we're iterating over the deps (as opposed to the outs) and therefore we have an idea of where the dep came from.

Another way of doing this would be to iterate using preorder_nodes only if the `--outs` option was enabled.

Reply:
But, code looks strange. If we show outs/deps then stages are edges, not pairs of stages. Why we iterate pairs then? Won't we miss something in the end?

Let's make some test, which will make it more obvious.

Reply:
I've changed it so that when we're showing outs, we're iterating over stages (which are the vertices/nodes in the graph).

I've also added a test.